File: packages/application/src/update-client.use-case.ts
import {
  Client,
  ClientNotFoundError,
  type IClientRepository,
  type ILogger,
} from '@monolegal/domain';

export interface UpdateClientInput {
  name?: string;
  email?: string;
}

export class UpdateClientUseCase {
  constructor(
    private readonly clientRepository: IClientRepository,
    private readonly logger: ILogger,
  ) {}

  async execute(id: string, input: UpdateClientInput): Promise<Client> {
    const existing = await this.clientRepository.findById(id);
    if (!existing) {
      throw new ClientNotFoundError(`Client not found: ${id}`);
    }

    const updatedProps = {
      name: input.name ?? existing.name,
      email: input.email ?? existing.email,
    };
    Client.create({ id: existing.id, ...updatedProps });

    const client = await this.clientRepository.update(id, updatedProps);
    this.logger.info('Client updated', { clientId: id });
    return client;
  }
}

File: backend/dotnet/src/Monolegal.Application/UpdateClientUseCase.cs
using Monolegal.Domain.Entities;
using Monolegal.Domain.Errors;
using Monolegal.Domain.Ports;

namespace Monolegal.Application;

public record UpdateClientInput(string? Name, string? Email);

public class UpdateClientUseCase(IClientRepository clientRepository, ILogger logger)
{
    public async Task<Client> ExecuteAsync(string id, UpdateClientInput input, CancellationToken cancellationToken = default)
    {
        var existing = await clientRepository.FindByIdAsync(id, cancellationToken);
        if (existing is null)
            throw new ClientNotFoundError($"Client not found: {id}");

        var name = input.Name ?? existing.Name;
        var email = input.Email ?? existing.Email;
        Client.Create(new ClientProps(existing.Id, name, email));

        var client = await clientRepository.UpdateAsync(id, name, email, cancellationToken);
        logger.Info("Client updated", new Dictionary<string, object?> { ["clientId"] = id });
        return client;
    }
}
